### PR TITLE
factory: Use lazy unmount

### PR DESCRIPTION
**Pull request by keloyang**
Fixes: #2918

Signed-off-by: Shukui Yang <keloyangsk@gmail.com>

### CODE REVIEW

**Comment by cmaf** [comment]
Hi @keloyang, could you update the commit with more details?

/test-ubuntu

**Review by fidencio — COMMENTED** [review]
@keloyang, I'm not convinced about this one.
Would you mind to elaborate the need of a lazy unmount here?

**Comment by keloyang** [comment]
@fidencio thanks for you review.  I have not found a  obvious problem with it, but I think it is better to use lazy umount, because it can reduce the probability of problems.

```
[root@centos0 test]# mkdir a b
[root@centos0 test]# mount -B a b
[root@centos0 test]# cd b
[root@centos0 b]# umount ../b
umount: /home/workspace/test/b: target is busy.
        (In some cases useful info about processes that use
         the device is found by lsof(8) or fuser(1))
[root@centos0 b]# umount ../b -l     # use lazy umount
[root@centos0 b]# 
```

**Comment by devimc** [comment]
thanks @keloyang , could you elaborate a little bit more about what issue are you facing? do we really need this in the factory?

cc @bergwolf @lifupan

**Comment by keloyang** [comment]
@devimc thanks for your review. also ping @bergwolf and @lifupan ptal.

we can have the following case,
1. start kata container with factory feature, this need kata-runtime config to enable factory and use initrd as base image.
2. start a kata container.
3. cd /root; cd /run/vc/vm/template dir, this will make /run/vc/vm/template to be in used. 
4. destroy vm template with kata-runtime factory destroy , and check the template mountpoint.
we can see  the template mountpoints will add everytime we repeat the above steps .
```
[root@centos1 template]# mount |grep template
[root@centos1 template]# docker run -ti --rm  --runtime untrusted-runtime --net none busybox echo

[root@centos1 template]# cd /root; cd /run/vc/vm/template/                
[root@centos1 template]# /kata/bin/kata-runtime factory destroy 
vm factory destroyed
[root@centos1 template]# mount |grep template
tmpfs on /run/vc/vm/template type tmpfs (rw,nosuid,nodev,relatime,seclabel,size=2105344k)
[root@centos1 template]# docker run -ti --rm  --runtime untrusted-runtime --net none busybox echo

[root@centos1 template]# cd /root; cd /run/vc/vm/template/      
[root@centos1 template]# /kata/bin/kata-runtime factory destroy 
vm factory destroyed
[root@centos1 template]# mount |grep template
tmpfs on /run/vc/vm/template type tmpfs (rw,nosuid,nodev,relatime,seclabel,size=2105344k)
tmpfs on /run/vc/vm/template type tmpfs (rw,nosuid,nodev,relatime,seclabel,size=2105344k)
```

**Comment by jodh-intel** [comment]
Thanks @keloyang. This does indeed look like a bug. I'd prefer we also make `template.close()` return an `error` as that can be passed back to the caller (or atleast logged somehow) in `template.prepareTemplateFiles()`.

@keloyang - Once @bergwolf / @lifupan have commented on this, would you also be able to raise a PR on Kata 2.0 (https://github.com/kata-containers/kata-containers/tree/2.0-dev/src/runtime)? Adding forward port label so we don't forget to do the port...

**Comment by lifupan** [comment]
Since the close() had ignored the error of umount, I think it's safe to do a lazy umount here.

**Comment by devimc** [comment]
thanks @keloyang - Could you update the commit message with the case that you mention?

> we can have the following case, .........

**Comment by keloyang** [comment]
@devimc I have updated the commit message. thanks for your review.

**Comment by jodh-intel** [comment]
> Since the close() had ignored the error of umount, I think it's safe to do a lazy umount here.

I agree, but strictly `close()` shouldn't have ignored the error so we could have found this problem quicker. But if we now use lazy umount we don't expect any errors, hence should really return `error` from `close()` "just in case" to avoid any future hidden bugs.

**Comment by jodh-intel** [comment]
added dnm label as there is still some debate around this.

**Comment by keloyang** [comment]
> I agree, but strictly close() shouldn't have ignored the error so we could have found this problem quicker. But if we now use lazy umount we don't expect any errors, hence should really return error from close() "just in case" to avoid any future hidden bugs.

@jodh-intel thanks for your review and thinking deeply
I agree with you, we shouldn't  ignored the error. but I think we still can use lazy umount. If we keep to use syscall.Unmount(*, 0),we can find the error and then we will looking for solutions. How to fix ? we can print err info, if we  find the error info later, we can  do a manual cleaning, alternatively we use lazy unmount, as far as my concerned,I like lazy unmount better, so I have add some log to close()， ptal.

File: virtcontainers/factory/template/template.go [hunk 1]
```diff
@@ -171,3 +178,10 @@ func (t *template) checkTemplateVM() error {
 	_, err = os.Stat(t.statePath + "/state")
 	return err
 }
+
+// Logger returns a logrus logger appropriate for logging template messages
+func (t *template) Logger() *logrus.Entry {
```
Comment (jodh-intel) [inline virtcontainers/factory/template/template.go:183]: Thanks for adding this. But can you "stitch" this into the existing logging facility by creating a `SetLogger()` function here and then calling it in:

- https://github.com/kata-containers/runtime/blob/master/virtcontainers/api.go#L49

Then, Having this logger function just return the global logger in this module that will now contain all the log fields. This is required for consistency and for https://github.com/kata-containers/tests/tree/master/cmd/log-parser#logfile-requirements.

**Review by jodh-intel — COMMENTED** [review]
Thanks for updating @keloyang!

Just one more tweak I think.

File: virtcontainers/factory/template/template.go (hunk 1 above)
Comment (keloyang) [reply virtcontainers/factory/template/template.go:183]: I have add SetLogger, ptal. thanks @jodh-intel

File: virtcontainers/factory/template/template.go [hunk 2]
```diff
@@ -171,3 +178,19 @@ func (t *template) checkTemplateVM() error {
...
+	return templateLog.WithFields(logrus.Fields{
+		"subsystem": "template",
+	})
+}
+
+// SetLogger sets the logger for the factory template.
+func SetLogger(ctx context.Context, logger logrus.FieldLogger) {
+	fields := logrus.Fields{
+		"source": "virtcontainers",
```
Comment (jodh-intel) [inline virtcontainers/factory/template/template.go:192]: Nit: Shouldn't this be:

```go
"source": "virtcontainers/factory/template",
```

**Review by jodh-intel — APPROVED** [review]
Thanks @keloyang.

File: virtcontainers/factory/template/template.go (hunk 2 above)
Comment (keloyang) [reply virtcontainers/factory/template/template.go]: I'm sorry, I  didn't understand it correctly. I have change it. @jodh-intel

**Comment by jodh-intel** [comment]
Thanks @keloyang.

**Comment by keloyang** [comment]
ping @lifupan @jodh-intel @fidencio

**Comment by fidencio** [comment]
@keloyang, we're aware of the PR and we'll move forward with this as soon as possible.
Right now, our metrics CI is failing and that's a blocker for pushing new PRs. Of course, we could bypass and do a git push directly instead of relying on GitHub "merge", but right now I don't see any issue with a priority that's high enough to do so.

So, all-in-all, this PR is on our radar and we'll get to this (and all the others) as soon as we unblock the queue.

**Comment by jodh-intel** [comment]
forward port PR: https://github.com/kata-containers/kata-containers/pull/939

**Pull request merged**